packages/playwright-core/src/server/trace/recorder/tracing.ts

```diff
@@ -131,14 +140,20 @@ export class Tracing implements InstrumentationListener, SnapshotterDelegate, Ha
       await fs.promises.appendFile(state.traceFile, JSON.stringify({ ...this._contextCreatedEvent, title: options.title, wallTime: Date.now() }) + '\n');
     });
 
-    this._context.instrumentation.addListener(this, this._context);
+    if (this._context)
+      this._context.instrumentation.addListener(this, this._context);
+    else
+      this._apiRequest.instrumentation.addListener(this, null);
```

Comment:
- Similarly here, it is not clear why we do not instrument apiRequest when we have a context, unless we have `context: BrowserContext | APIRequestContext` instead. WDYT?
- I don't think passing `null` works as you'd expect. It means "listen to all events from all contexts", so you'll get all the instrumentation in the world. Instead, we should be able to filter by `BrowserContext | APIRequestContext` instead.

Reply:
Updated, ptal.